pkg/blockbuilder/tsdb.go

```diff
@@ -0,0 +1,393 @@
+		return nil, err
+	}
+
+	userID := tenant.tenantID
+	userLogger := util_log.WithUserID(userID, b.logger)
+
+	udb := &userTSDB{
+		userID: userID,
+	}
+
+	db, err := tsdb.Open(udir, userLogger, nil, &tsdb.Options{
+		RetentionDuration:           0,
+		MinBlockDuration:            2 * time.Hour.Milliseconds(),
+		MaxBlockDuration:            2 * time.Hour.Milliseconds(),
+		NoLockfile:                  true,
+		StripeSize:                  b.config.TSDB.StripeSize,
+		HeadChunksWriteBufferSize:   b.config.TSDB.HeadChunksWriteBufferSize,
+		HeadChunksWriteQueueSize:    b.config.TSDB.HeadChunksWriteQueueSize,
+		WALSegmentSize:              -1, // No WAL
+		SeriesLifecycleCallback:     udb,
+		BlocksToDelete:              udb.blocksToDelete,
+		IsolationDisabled:           true,
+		EnableOverlappingCompaction: false,                                                // always false since Mimir only uploads lvl 1 compacted blocks
+		OutOfOrderTimeWindow:        b.limits.OutOfOrderTimeWindow(userID).Milliseconds(), // The unit must be same as our timestamps.
+		OutOfOrderCapMax:            int64(b.config.TSDB.OutOfOrderCapacityMax),
+		EnableNativeHistograms:      b.limits.NativeHistogramsIngestionEnabled(userID),
+		SecondaryHashFunction:       nil, // TODO(codesome): May needed when applying limits. Used to determine the owned series by an ingesters
+	}, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	db.DisableCompactions()
+
+	udb.DB = db
+
+	return udb, nil
+}
+
+// Function to upload the blocks.
+type blockUploader func(_ context.Context, tenantID, dbDir string, blockIDs []string) error
+
+// CompactAndUpload compacts the blocks of all the TSDBs and uploads them.
+// uploadBlocks is a function that uploads the blocks to the required storage.
+// If the functions returns an error, the DBs that are not successfully compacted or
+// uploaded will not be cleared from the TSDBBuilder.
+func (b *TSDBBuilder) CompactAndUpload(ctx context.Context, uploadBlocks blockUploader) (int, error) {
+	b.tsdbsMu.Lock()
+	defer b.tsdbsMu.Unlock()
+
+	level.Info(b.logger).Log("msg", "compacting and uploading blocks", "num_tsdb", len(b.tsdbs))
+
+	if len(b.tsdbs) == 0 {
+		return 0, nil
+	}
+
+	eg, ctx := errgroup.WithContext(ctx)
+	if b.config.TSDB.ShipConcurrency > 0 {
+		eg.SetLimit(b.config.TSDB.ShipConcurrency)
+	}
+	var numBlocks atomic.Int64
+	for tenant, db := range b.tsdbs {
+		// Change scope of for loop variables.
+		tenant := tenant
+		db := db
+
+		eg.Go(func() error {
+			if err := db.compactEverything(ctx); err != nil {
+				return err
+			}
+
+			var blockIDs []string
+			dbDir := db.Dir()
+			for _, b := range db.Blocks() {
+				blockIDs = append(blockIDs, b.Meta().ULID.String())
+			}
+			numBlocks.Add(int64(len(blockIDs)))
+
+			if err := db.Close(); err != nil {
+				return err
+			}
+
+			if err := uploadBlocks(ctx, tenant.tenantID, dbDir, blockIDs); err != nil {
+				return err
+			}
+
+			delete(b.tsdbs, tenant)
```

Comment:
I just realised there's a race here: `t.tsdbs` is manipulated in `eg.Go()` which means there could be multiple goroutines manipulating it.

Reply:
Race should be fixed now